// website-carbon-calculator/build/main/lib/websiteCarbonCalculator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebsiteCarbonCalculator = void 0;
const node_fetch_1 = __importDefault(require("node-fetch"));
const constants_1 = require("./constants");
const websiteCarbonCalculatorError_1 = require("./websiteCarbonCalculatorError");
/**
 * Based on https://gitlab.com/wholegrain/carbon-api-2-0/-/tree/master/
 */
class WebsiteCarbonCalculator {
    constructor({ pagespeedApiKey }) {
        if (!pagespeedApiKey || !pagespeedApiKey.length) {
            throw new websiteCarbonCalculatorError_1.WebsiteCarbonCalculatorError("Please set the PageSpeed API Key. (i.e: new WebsiteCarbonCalculator({pagespeedApiKey: 'YOU_API_KEY'}))");
        }
        this.PAGE_SPEED_API_KEY = pagespeedApiKey;
    }
    getNetworkTraffic({ url, key, }) {
        return node_fetch_1.default(`${constants_1.GOOGLE_PAGESPEED_API_ENDPOINT}?url=${url}&key=${key}`).then((res) => res.json());
    }
    getGreenWeb(url) {
        const parsedURL = new URL(url);
        return node_fetch_1.default(`${constants_1.GREEN_FOUNDATION_API_ENDPOINT}/${parsedURL.host}`).then((res) => res.json());
    }
    async calculateByURL(url) {
        var _a, _b, _c;
        const normalizedURL = this.normalizeURL(url);
        const [pagespeedapi, greenweb] = await Promise.all([
            this.getNetworkTraffic({
                url: normalizedURL,
                key: this.PAGE_SPEED_API_KEY,
            }),
            this.getGreenWeb(normalizedURL),
        ]);
        if (pagespeedapi === null || pagespeedapi === void 0 ? void 0 : pagespeedapi.error) {
            throw new websiteCarbonCalculatorError_1.WebsiteCarbonCalculatorError(pagespeedapi.error.message);
        }
        const isGreenHost = greenweb === null || greenweb === void 0 ? void 0 : greenweb.green;
        if (!((_c = (_b = (_a = pagespeedapi === null || pagespeedapi === void 0 ? void 0 : pagespeedapi.lighthouseResult) === null || _a === void 0 ? void 0 : _a.audits['network-requests']) === null || _b === void 0 ? void 0 : _b.details) === null || _c === void 0 ? void 0 : _c.items)) {
            throw new websiteCarbonCalculatorError_1.WebsiteCarbonCalculatorError("Sorry, traffic couldn't be estimated.");
        }
        const bytesTransferred = this.calculateTransferredBytes(pagespeedapi.lighthouseResult.audits['network-requests'].details.items);
        // Calculate the statistics as we need the co2 emissions
        const statistics = this.getStatistics(bytesTransferred);
        // pull the co2 relative to the energy
        const co2 = isGreenHost === true
            ? statistics['co2']['renewable']['grams']
            : statistics['co2']['grid']['grams'];
        return {
            url: normalizedURL,
            bytesTransferred,
            isGreenHost: isGreenHost,
            co2PerPageview: co2,
        };
    }
    calculateTransferredBytes(items) {
        return items
            .filter((item) => item.transferSize !== undefined)
            .map((item) => item.transferSize)
            .reduce((prev, next) => {
            return prev + next;
        }, 0);
    }
    getStatistics(bytes) {
        const bytesAdjusted = this.adjustDataTransfer(bytes);
        const energy = this.energyConsumption(bytesAdjusted);
        const co2Grid = this.getCo2Grid(energy);
        const co2Renewable = this.getCo2Renewable(energy);
        return {
            adjustedBytes: bytesAdjusted,
            energy: energy,
            co2: {
                grid: {
                    grams: co2Grid,
                    litres: this.co2ToLitres(co2Grid),
                },
                renewable: {
                    grams: co2Renewable,
                    litres: this.co2ToLitres(co2Renewable),
                },
            },
        };
    }
    normalizeURL(url) {
        try {
            const normalizedURL = new URL(url);
            return `${normalizedURL.protocol}//${normalizedURL.hostname}`;
        }
        catch (_) {
            throw new websiteCarbonCalculatorError_1.WebsiteCarbonCalculatorError('Ops! This is an invalid URL.');
        }
    }
    adjustDataTransfer(val) {
        return (val * constants_1.RETURNING_VISITOR_PERCENTAGE +
            constants_1.PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD *
                val *
                constants_1.FIRST_TIME_VIEWING_PERCENTAGE);
    }
    energyConsumption(bytes) {
        return bytes * (constants_1.KWG_PER_GB / 1073741824);
    }
    getCo2Grid(energy) {
        return energy * constants_1.CARBON_PER_KWG_GRID;
    }
    getCo2Renewable(energy) {
        return (energy * constants_1.PERCENTAGE_OF_ENERGY_IN_DATACENTER * constants_1.CARBON_PER_KWG_RENEWABLE +
            energy *
                constants_1.PERCENTAGE_OF_ENERGY_IN_TRANSMISSION_AND_END_USER *
                constants_1.CARBON_PER_KWG_GRID);
    }
    co2ToLitres(co2) {
        return co2 * constants_1.CO2_GRAMS_TO_LITRES;
    }
}
exports.WebsiteCarbonCalculator = WebsiteCarbonCalculator;
