// website-carbon-calculator/build/main/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GREEN_FOUNDATION_API_ENDPOINT = exports.GOOGLE_PAGESPEED_API_ENDPOINT = exports.CO2_GRAMS_TO_LITRES = exports.PERCENTAGE_OF_ENERGY_IN_TRANSMISSION_AND_END_USER = exports.PERCENTAGE_OF_ENERGY_IN_DATACENTER = exports.CARBON_PER_KWG_RENEWABLE = exports.CARBON_PER_KWG_GRID = exports.PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD = exports.FIRST_TIME_VIEWING_PERCENTAGE = exports.RETURNING_VISITOR_PERCENTAGE = exports.KWG_PER_GB = void 0;
exports.KWG_PER_GB = 1.805;
exports.RETURNING_VISITOR_PERCENTAGE = 0.75;
exports.FIRST_TIME_VIEWING_PERCENTAGE = 0.25;
exports.PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD = 0.02;
exports.CARBON_PER_KWG_GRID = 475;
exports.CARBON_PER_KWG_RENEWABLE = 33.4;
exports.PERCENTAGE_OF_ENERGY_IN_DATACENTER = 0.1008;
exports.PERCENTAGE_OF_ENERGY_IN_TRANSMISSION_AND_END_USER = 0.8992;
exports.CO2_GRAMS_TO_LITRES = 0.5562;
exports.GOOGLE_PAGESPEED_API_ENDPOINT = 'https://www.googleapis.com/pagespeedonline/v5/runPagespeed';
exports.GREEN_FOUNDATION_API_ENDPOINT = 'https://api.thegreenwebfoundation.org/greencheck';
