// website-carbon-calculator/build/main/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Group = exports.Protocol = exports.ScoreDisplayMode = exports.Type = void 0;
var Type;
(function (Type) {
    Type["Opportunity"] = "opportunity";
    Type["Table"] = "table";
})(Type = exports.Type || (exports.Type = {}));
var ScoreDisplayMode;
(function (ScoreDisplayMode) {
    ScoreDisplayMode["Binary"] = "binary";
    ScoreDisplayMode["Informative"] = "informative";
    ScoreDisplayMode["NotApplicable"] = "notApplicable";
    ScoreDisplayMode["Numeric"] = "numeric";
})(ScoreDisplayMode = exports.ScoreDisplayMode || (exports.ScoreDisplayMode = {}));
var Protocol;
(function (Protocol) {
    Protocol["H2"] = "h2";
})(Protocol = exports.Protocol || (exports.Protocol = {}));
var Group;
(function (Group) {
    Group["Budgets"] = "budgets";
    Group["Diagnostics"] = "diagnostics";
    Group["LoadOpportunities"] = "load-opportunities";
    Group["Metrics"] = "metrics";
})(Group = exports.Group || (exports.Group = {}));
